### PR TITLE
[Fluid-SDEM] Including new update subscale algorithm

### PR DESCRIPTION
**Pull request by jgonzalezusua**
**📝 Description**
This PR is aimed to include some developments in the fluid elements used to couple with SDEM `d_vms_dem_coupled.cpp`, `qs_vms_dem_coupled.cpp`, `alternative_d_vms_dem_coupled.cpp` and `alternative_qs_vms_dem_coupled.cpp`. Those developments are related to the inclusion of the `FullProjection()` method in the bdf2 scheme when the DEM coupling is used. Finally, we introduced a relaxed version of Newton-Raphson strategy in the `SwimmingDEMApplication` called `relaxed_residualbased_newton_raphson_strategy.h`.

Please mark the PR with appropriate tags: 
- Applications

**🆕 Changelog**

-  Including `relaxed_residualbased_newton_raphson_strategy.h` in which a relaxation parameter is introduced.
- Adding `FullProjection()` to the `bdf2_turbulent_schemeDEMCoupled.h` and adding its support to the fluid dem coupled elements 
- Adding some manufactured solutions to `SwimmingDEMApplication`
- Adding more print variables to the `hdf5_script.py`.

### CODE REVIEW

File: applications/FluidDynamicsApplication/custom_elements/d_vms_dem_coupled.cpp [hunk 1]
```diff
@@ -489,6 +527,65 @@ void DVMSDEMCoupled<TElementData>::AddMassStabilization(
...
+    GeometryType& r_geometry = this->GetGeometry();
+    unsigned int Row = 0;
+    for (SizeType i = 0; i < NumNodes; ++i)
+    {
+        r_geometry[i].SetLock(); // So it is safe to write in the node in OpenMP
+        array_1d<double,3>& rMomValue = r_geometry[i].FastGetSolutionStepValue(ADVPROJ);
+        for (unsigned int d = 0; d < Dim; ++d)
+            rMomValue[d] += MomentumRHS[Row++];
+        r_geometry[i].FastGetSolutionStepValue(DIVPROJ) += MassRHS[i];
+        r_geometry[i].FastGetSolutionStepValue(NODAL_AREA) += NodalArea[i];
```
Comment (GuillermoCasas) [inline applications/FluidDynamicsApplication/custom_elements/d_vms_dem_coupled.cpp:584]: Is it necessary to calculate this at every iteration? Should this not be calculated outside the element?

File: applications/SwimmingDEMApplication/custom_strategies/strategies/bdf2_turbulent_schemeDEMCoupled.h [hunk 2]
```diff
@@ -139,8 +139,158 @@ class BDF2TurbulentSchemeDEMCoupled : public BDF2TurbulentScheme<TSparseSpace, T
...
+
+            // Update iteration variables
+#pragma omp parallel for
+            for (int i_node = 0; i_node < n_nodes; ++i_node) {
+                auto ind = rModelPart.NodesBegin() + i_node;
+                const double Area = ind->FastGetSolutionStepValue(NODAL_AREA); // Ml dx = b - Mc x
+                dMomProj = ind->GetValue(ADVPROJ) / Area;
+                dMassProj = ind->GetValue(DIVPROJ) / Area;
+
+                RelMomErr += sqrt( dMomProj[0]*dMomProj[0] + dMomProj[1]*dMomProj[1] + dMomProj[2]*dMomProj[2]);
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_strategies/strategies/bdf2_turbulent_schemeDEMCoupled.h]: std::abs, std::sqrt...

File: applications/SwimmingDEMApplication/python_scripts/fluid_DEM_coupling_solver.py [hunk 3]
```diff
@@ -15,6 +15,13 @@ class FluidDEMSolver(FluidSolver):
 
     ## FluidDEMSolver specific methods.
 
+    def _TimeBufferIsInitialized(self):
+        # We always have one extra old step if we are not using a manufactured solution (step 0, read from input)
+        if self.main_model_part.ProcessInfo[KratosSDEM.MANUFACTURED]:
+            return self.main_model_part.ProcessInfo[KratosMultiphysics.STEP] + 2 >= self.GetMinimumBufferSize()
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/python_scripts/fluid_DEM_coupling_solver.py:21]: this is hacky... Let us see how we can improve this in the future...

File: applications/SwimmingDEMApplication/python_scripts/hdf5_script.py [hunk 4]
```diff
@@ -26,36 +26,42 @@ def __init__(self, test_number):
...
         self.av_mod_error = []
-        self.mean_iteration = []
+        self.n_iterations = []
         self.problem_path = os.getcwd()
         self.file_path = os.path.join(str(self.problem_path),self.parameters["file_name"].GetString())
         self.dtype = np.float64
         self.group_name = str(test_number)
 
-    def WriteData(self, error_model_part, velocity_error_projected, pressure_error_projected, projection_type, model_type, subscale_type):
+    def WriteData(self, error_model_part, velocity_error_projected, pressure_error_projected, projection_type, model_type, subscale_type, reynolds_number, porosity_mean, n_iterations, max_iteration, relax_alpha):
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/python_scripts/hdf5_script.py]: Parameters into different lines or better: think about how to encapsulate them

**Review by GuillermoCasas — CHANGES_REQUESTED** [review]
Just make the small changes and I will approve

File: applications/FluidDynamicsApplication/custom_elements/d_vms_dem_coupled.cpp (hunk 1 above)
Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/custom_elements/d_vms_dem_coupled.cpp:584]: As the `bdf2_turbulent_scheme.h` reinitializes the `NODAL_AREA` just before calling this method, we need to calculate the variable inside the element.

File: applications/SwimmingDEMApplication/custom_strategies/strategies/bdf2_turbulent_schemeDEMCoupled.h (hunk 2 above)
Comment (jgonzalezusua) [reply applications/SwimmingDEMApplication/custom_strategies/strategies/bdf2_turbulent_schemeDEMCoupled.h]: Ok

**Comment by jgonzalezusua** [comment]
The small requested changes have been addressed. The improvements proposed in the review are being implemented in another branch

**Review by GuillermoCasas — APPROVED** [review]

**Pull request merged**